File: 0000-0999/0092.ReverseLinkedListII.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        if not head:
            return head
        dummy = pre = ListNode(0)
        dummy.next = head
        for i in xrange(m-1):
            pre = pre.next
        start = pre.next
        then = start.next
        for i in xrange(m, n):
            start.next = then.next
            then.next = pre.next
            pre.next = then
            then = start.next
        return dummy.next

File: 0000-0999/0092.ReverseLinkedListII.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, left int, right int) *ListNode {
    dummy := &ListNode{Next: head}
    pre := dummy
    for i := 1; i < left; i++ {
        pre = pre.Next
    }
    l, r := pre.Next, pre.Next
    for i := left; i < right; i++ {
        r = r.Next
    }
    tail := r.Next
    r.Next = nil
    for l != nil {
        tmp := l.Next
        l.Next = tail
        tail = l
        l = tmp
    }
    pre.Next = r
    return dummy.Next
}